academy/0.0-B9lab-Blockchains/5_crypto.md

```diff
@@ -0,0 +1,180 @@
+
+Example (Linux):
+
+```
+// Create SECP256K1 private key with explicit parameters for backward compatibility
+$ openssl ecparam -name secp256k1 -genkey -noout -out secp256k1-key.pem -param_enc explicit
+// Create public key
+$ openssl ec -in secp256k1-key.pem -pubout -out secp256k1-key-pub.pem
+// Show public key
+$ openssl ec -in secp256k1-key-pub.pem -pubin -text -noout
+
+// Create RSA private key
+$ openssl genrsa -des3 -out rsa-key.pem 2048
+Generating RSA private key, 2048 bit long modulus
+.....................................................+++
+...........+++
+e is 65537 (0x10001)
+Enter pass phrase for rsa-key.pem:
+Verifying - Enter pass phrase for rsa-key.pem:
+// Create public key
+$ openssl rsa -in rsa-key.pem -outform PEM -pubout -out rsa-key-pub.pem
+Enter pass phrase for rsa-key.pem:
+writing RSA key
+```
+
+This is like a password that is used to encrypt your private key on disk. If the private key was not encrypted, it would be at greater risk of theft. Since you are just testing here, you can put nothing or a simple word. But remember that whenever you create keys in the future, you need to protect them with a proper password.
+
+Note that you may need openssl version 1.0 or newer.
+
+#### Encrypt and decrypt
+
+Alice wants to send a message to Bob, and for Bob's eyes only:
+
+* Bob gives Alice his public key
+* Alice uses Bob's public key to encrypt the message
+* Alice sends Bob the encrypted message
+* Bob decrypts the message with his private key
+
+![](images/00_11_rsa_keys_v1.png)
+
+Example:
+
+```bash
+// Encrypt file
+$ openssl pkeyutl -encrypt -pubin -inkey rsa-key-pub.pem -in helloworld.txt -out helloworld.enc
+// Decrypt file
+$ openssl pkeyutl -decrypt -inkey rsa-key.pem -in helloworld.enc -out helloworld2.txt
+```
+
+If you receive an error, try with `openssl rsautl` instead.
+
+#### Sign and verify
+
+Alice wants to make sure that Bob's public announcement is indeed from Bob:
+
+* Bob gives Alice his public key
+* Bob signs his announcement with his private key
+* Bob sends Alice his announcement and its signature
+* Alice verifies the signature with Bob's public key
+
+![](images/00_12_digital_signature_keys_v2.png)
+
+Example:
+
+```bash
+// Sign file hash
+$ openssl dgst -sha256 -sign secp256k1-key.pem -out helloworld-bin.sha256 helloworld.txt
+// Encode signature in Base64
+$ openssl base64 -in helloworld-bin.sha256 -out helloworld.sha256
+
+// Decode signature form Base64
+$ openssl base64 -d -in helloworld.sha256 -out helloworld-bin-decoded.sha256
+// Verify signature
+$ openssl dgst -sha256 -verify secp256k1-key-pub.pem -signature helloworld-bin-decoded.sha256 helloworld.txt
+Verified OK
+```
+
+#### Mix and match
+
+It is possible to mix both ideas, whereby Alice encrypts her message with Bob's public key, then signs the encrypted file with her private key. Upon reception, Bob verifies the signature with Alice's public key, to be sure it came from Alice, then decrypts the file with his private key.
```

Comment:
I made a list out of it to be more consistent in regard to the two previous scenario examples.